sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -628,6 +576,73 @@ class GpuDynamicPartitionDataSingleWriter(
     }
   }
 
+  /**
+   * Write a Batch.
+   *
+   * Note: The `batch` will be closed in this function.
+   *
+   * @param batch the batch to be written
+   * @param maxRecordsPerFile the max number of rows per file
+   * @param partPath the partition directory
+   */
+  private def writeBatch(batch: ColumnarBatch, maxRecordsPerFile: Long, partPath: String) = {
+    val (table, dataTypes) = withResource(batch) { _ =>
+      (GpuColumnVector.from(batch), GpuColumnVector.extractTypes(batch))
+    }
+    writeTable(table, dataTypes, maxRecordsPerFile, partPath)
+  }
+
+
+  /**
+   * Write a Table.
+   *
+   * Note: The `table` will be closed in this function.
+   *
+   * @param table the table to be written
+   * @param outDataTypes the data types of the table
+   * @param maxRecordsPerFile the max number of rows per file
+   * @param partPath the partition directory
+   */
+  private def writeTable(
+    table: Table, outDataTypes: Array[DataType], maxRecordsPerFile: Long, partPath: String) = {
+    val splitIndexes = getSplitIndexes(
+      maxRecordsPerFile,
+      currentWriterStatus.recordsInFile,
+      table.getRowCount()
+    )
+    var needNewWriter = currentWriterStatus.recordsInFile >= maxRecordsPerFile
+
+    val tabs = withResource(table) { _ => table.contiguousSplit(splitIndexes: _*)}
+    withResource(tabs) { _ =>
+      tabs.foreach(b => {
```

Comment:
Not new with this PR, but we're leaving the other unspillable contiguous splits in GPU memory when we write the current split which releases the GPU semaphore.  Filed #6980 to track that problem.